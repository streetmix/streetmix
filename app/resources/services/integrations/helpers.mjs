import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
