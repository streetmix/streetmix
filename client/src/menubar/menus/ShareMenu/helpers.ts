import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
