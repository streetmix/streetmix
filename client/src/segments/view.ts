import { percentToNumber } from '@streetmix/utils'

import { images } from '../app/load_resources'
import { formatMessage } from '../locales/locale'
import { saveStreetToServerIfNecessary } from '../streets/data_model'
import { recalculateWidth } from '../streets/width'
import store from '../store'
import { updateSegments, changeSegmentProperties } from '../store/slices/street'
import { getSegmentInfo, getSegmentVariantInfo, getSpriteDef } from './info'
import { drawScatteredSprites } from './scatter'
import {
  TILE_SIZE,
  TILESET_POINT_PER_PIXEL,
  TILE_SIZE_ACTUAL,
  MAX_SEGMENT_LABEL_LENGTH,
  BUILDING_LEFT_POSITION,
  BUILDING_RIGHT_POSITION
} from './constants'
import PEOPLE from './people.yaml'

import type {
  VariantInfo,
  VariantInfoDimensions,
  Segment,
  BoundaryPosition,
  UnknownVariantInfo
} from '@streetmix/types'

// Adjust spacing between people to be slightly closer
const PERSON_SPACING_ADJUSTMENT = -0.1 // in meters
const PERSON_SPRITE_OFFSET_Y = 10 // in pixels

/**
 * Draws SVG sprite to canvas
 *
 * @param id - identifier of sprite
 * @param ctx
 * @param sx - x position of sprite to read from (default = 0)
 * @param sy - y position of sprite to read from (default = 0)
 * @param sw - sub-rectangle width to draw
 * @param sh - sub-rectangle height to draw
 * @param dx - x position on canvas
 * @param dy - y position on canvas
 * @param dw - destination width to draw
 * @param dh - destination height to draw
 * @param multiplier - scale to draw at (default = 1)
 * @param dpi
 */
export function drawSegmentImage (
  id: string,
  ctx: CanvasRenderingContext2D,
  sx: number = 0,
  sy: number = 0,
  sw: number | undefined,
  sh: number | undefined,
  dx: number,
  dy: number,
  dw: number | undefined,
  dh: number | undefined,
  multiplier: number = 1,
  dpi?: number
): void {
  // If asked to render a source or destination image with width or height
  // that is equal to or less than 0, bail. Attempting to render such an image
  // will throw an IndexSizeError error in Firefox.
  // If any of these values are undefined, the comparison evaluates to false.
  // So it's valid, even though it's a type error. Overridden with `!`
  if (sw! <= 0 || sh! <= 0 || dw! <= 0 || dh! <= 0) return

  // Settings
  const state = store.getState()
  dpi = dpi || state.system.devicePixelRatio || 1
  const debugRect = state.flags.DEBUG_SEGMENT_CANVAS_RECTANGLES.value || false

  // Get image definition
  const svg = images.get(id)
  // Source width and height is based off of intrinsic image width and height,
  // but it can be overridden in the parameters, e.g. when repeating sprites
  // in a sequence and the last sprite needs to be truncated
  sw = sw === undefined ? (svg.width as number) : sw * TILESET_POINT_PER_PIXEL
  sh = sh === undefined ? (svg.height as number) : sh * TILESET_POINT_PER_PIXEL

  // We can't read `.naturalWidth` and `.naturalHeight` properties from
  // the image in IE11, which returns 0. This is why width and height are
  // stored as properties from when the image is first cached
  // All images are drawn at 2x pixel dimensions so divide in half to get
  // actual width / height value then multiply by system pixel density
  //
  // dw/dh (and later sw/sh) can be 0, so don't use falsy checks
  dw = dw === undefined ? (svg.width as number) / TILESET_POINT_PER_PIXEL : dw
  dh = dh === undefined ? (svg.height as number) / TILESET_POINT_PER_PIXEL : dh
  dw *= multiplier * dpi
  dh *= multiplier * dpi

  // Set render dimensions based on pixel density
  dx *= dpi
  dy *= dpi

  // These rectangles are telling us that we're drawing at the right places.
  if (debugRect === true) {
    ctx.fillStyle = 'rgba(0, 0, 0, 0.2)'
    ctx.fillRect(dx, dy, dw, dh)
  }

  // Round all values to whole numbers when we render. Decimal values can
  // introduce tiny "seams" at tiled assets, for example.
  ctx.drawImage(
    svg.img,
    Math.round(sx),
    Math.round(sy),
    Math.round(sw),
    Math.round(sh),
    Math.round(dx),
    Math.round(dy),
    // destination sprites round up to prevent gaps in tiled sprites
    Math.ceil(dw),
    Math.ceil(dh)
  )
}

/**
 * When rendering a stack of sprites, sometimes the resulting image will extend
 * beyond the left or right edge of the segment's width. (For instance, a
 * "tree" segment might be 0.5m wide, but the actual width of the tree sprite
 * will need more than 0.5m width to render.) This calculates the actual left,
 * right, and center Y-values needed to render sprites so that they are not
 * truncated at the edge of the segment.
 *
 * @param variantInfo - segment variant info
 * @param actualWidth - segment's actual real life width
 */
export function getVariantInfoDimensions (
  variantInfo: VariantInfo | UnknownVariantInfo,
  actualWidth: number = 0
): VariantInfoDimensions {
  // Convert actualWidth to units that work with images' intrinsic dimensions
  const displayWidth = actualWidth * TILE_SIZE_ACTUAL

  const center = displayWidth / 2
  let left = center
  let right = center
  let newLeft, newRight

  const graphics = variantInfo.graphics

  if (graphics.center) {
    const sprites = Array.isArray(graphics.center)
      ? graphics.center
      : [graphics.center]

    for (let l = 0; l < sprites.length; l++) {
      const sprite = getSpriteDef(sprites[l])
      const svg = images.get(sprite.id)

      // If svg is missing, let it not affect this calculation
      if (svg) {
        // TODO: This doesn't take into account % offsets
        newLeft = center - svg.width / 2 - (sprite.offsetX ?? 0)
        newRight = center + svg.width / 2 + (sprite.offsetX ?? 0)

        if (newLeft < left) {
          left = newLeft
        }
        if (newRight > right) {
          right = newRight
        }
      }
    }
  }

  if (graphics.left) {
    const sprites = Array.isArray(graphics.left)
      ? graphics.left
      : [graphics.left]

    for (let l = 0; l < sprites.length; l++) {
      const sprite = getSpriteDef(sprites[l])
      const svg = images.get(sprite.id)

      if (svg) {
        newLeft = sprite.offsetX ?? 0
        newRight = svg.width + (sprite.offsetX ?? 0)

        if (newLeft < left) {
          left = newLeft
        }
        if (newRight > right) {
          right = newRight
        }
      }
    }
  }

  if (graphics.right) {
    const sprites = Array.isArray(graphics.right)
      ? graphics.right
      : [graphics.right]

    for (let l = 0; l < sprites.length; l++) {
      const sprite = getSpriteDef(sprites[l])
      const svg = images.get(sprite.id)

      if (svg) {
        newLeft = displayWidth - (sprite.offsetX ?? 0) - svg.width
        newRight = displayWidth - (sprite.offsetX ?? 0)

        if (newLeft < left) {
          left = newLeft
        }
        if (newRight > right) {
          right = newRight
        }
      }
    }
  }

  if (graphics.repeat && graphics.repeat[0]) {
    newLeft = center - displayWidth / 2
    newRight = center + displayWidth / 2

    if (newLeft < left) {
      left = newLeft
    }
    if (newRight > right) {
      right = newRight
    }
  }

  // If a segment has a "minimum renderable width", the left and right values
  // are overridden based on the `quirks.minWidth` property. This prevents
  // left/right assets from rendering in strange positions.
  if (graphics.quirks?.minWidth) {
    const minDimension = graphics.quirks.minWidth * TILE_SIZE_ACTUAL
    if (displayWidth < minDimension) {
      left = (displayWidth - minDimension) / 2
      right = (minDimension - displayWidth) / 2 + displayWidth
    }
  }

  return {
    left: left / TILE_SIZE_ACTUAL,
    right: right / TILE_SIZE_ACTUAL,
    center: center / TILE_SIZE_ACTUAL
  }
}

/**
 * Given an elevation value (in meters), return the pixel height it should be
 * rendered at.
 *
 * @param elevation
 * @returns pixels
 */
export function getElevation (elevation: number): number {
  return elevation * TILE_SIZE
}

/**
 *
 * @param ctx
 * @param type
 * @param variantString
 * @param actualWidth - The real-world width of a segment, in meters
 * @param offsetLeft
 * @param groundBaseline
 * @param randSeed
 * @param multiplier
 * @param dpi
 */
export function drawSegmentContents (
  ctx: CanvasRenderingContext2D,
  type: string,
  variantString: string,
  actualWidth: number,
  offsetLeft: number,
  groundBaseline: number,
  elevation: number = 0,
  randSeed: string,
  multiplier: number,
  dpi: number
): void {
  const variantInfo = getSegmentVariantInfo(type, variantString)
  const graphics = variantInfo.graphics

  // TODO: refactor this variable
  const segmentWidth = actualWidth * TILE_SIZE

  const dimensions = getVariantInfoDimensions(variantInfo, actualWidth)
  const left = dimensions.left
  const minWidthQuirk = graphics.quirks?.minWidth

  const groundLevelOffsetY = variantInfo.offsetY ?? 0
  const elevationValue = getElevation(elevation)
  const groundLevel =
    groundBaseline -
    multiplier * (groundLevelOffsetY / TILESET_POINT_PER_PIXEL) -
    multiplier * elevationValue

  const coastmixMode = store.getState().flags.COASTMIX_MODE.value

  if (graphics.repeat) {
    // Convert single string or object values to single-item array
    let sprites = Array.isArray(graphics.repeat)
      ? graphics.repeat
      : [graphics.repeat]
    // Convert array of strings into array of objects
    // If already an object, pass through
    sprites = sprites.map((def) =>
      typeof def === 'string' ? { id: def } : def
    )

    for (let l = 0; l < sprites.length; l++) {
      const sprite = getSpriteDef(sprites[l].id)
      const svg = images.get(sprite.id)

      // Skip drawing if sprite is missing
      if (!svg) continue

      let width = (svg.width / TILE_SIZE_ACTUAL) * TILE_SIZE
      const padding = sprites[l].padding ?? 0

      let drawWidth
      // If quirks.minWidth is defined, and the segment width is less than that
      // value, then the draw width is the minimum renderable width, not the
      // segment width. Skip this for ground assets.
      if (
        minWidthQuirk &&
        actualWidth < minWidthQuirk &&
        !sprite.id.includes('ground')
      ) {
        drawWidth = minWidthQuirk * TILE_SIZE - padding * 2 * TILE_SIZE
      } else {
        drawWidth = segmentWidth - padding * 2 * TILE_SIZE
      }

      const countX = Math.floor(drawWidth / width) + 1

      let repeatStartX
      if (left < 0) {
        // If quirks.minWidth is defined, and the segment width is less than
        // that value, then render the left edge at minimum width boundary,
        // don't reposition it along the segment's left edge. Skip this process
        // if it's a ground asset.
        if (
          minWidthQuirk &&
          actualWidth < minWidthQuirk &&
          !sprite.id.includes('ground')
        ) {
          repeatStartX = 0
        } else {
          // This is for rendering beyond the left edge of the segment
          repeatStartX = -left * TILE_SIZE
        }
      } else {
        // This is the normal render logic.
        repeatStartX = 0
      }

      // The distance between the top of the sprite and the ground is
      // calculated by subtracting the height of the sprite with the # of
      // pixels to get to the point of the sprite which should align with
      // the ground.
      const distanceFromGround =
        multiplier *
        TILE_SIZE *
        ((svg.height - (sprite.originY ?? 0) + (sprite.offsetY ?? 0)) /
          TILE_SIZE_ACTUAL)

      // Right now only ground items repeat in the Y direction
      const height = (svg.height / TILE_SIZE_ACTUAL) * TILE_SIZE

      // countY should always be at minimum 1.
      const countY = sprite.id.startsWith('ground--')
        ? Math.ceil((ctx.canvas.height / dpi - groundLevel) / height)
        : 1

      for (let i = 0; i < countX; i++) {
        // remainder
        if (i === countX - 1) {
          width = drawWidth - (countX - 1) * width
        }

        for (let j = 0; j < countY; j++) {
          // If the sprite being rendered is the ground, dy is equal to the
          // groundLevel. If not, dy is equal to the groundLevel minus the
          // distance the sprite will be from the ground.
          drawSegmentImage(
            sprite.id,
            ctx,
            undefined,
            undefined,
            width,
            undefined,
            offsetLeft +
              padding * TILE_SIZE * multiplier +
              (repeatStartX + i * (svg.width / TILE_SIZE_ACTUAL) * TILE_SIZE) *
                multiplier,
            sprite.id.startsWith('ground--')
              ? groundLevel + height * j
              : groundLevel - distanceFromGround,
            width,
            undefined,
            multiplier,
            dpi
          )
        }
      }
    }
  }

  if (graphics.left) {
    const sprites = Array.isArray(graphics.left)
      ? graphics.left
      : [graphics.left]
    for (let l = 0; l < sprites.length; l++) {
      const sprite = getSpriteDef(sprites[l])
      const svg = images.get(sprite.id)

      // Skip drawing if sprite is missing
      if (!svg) continue

      let x
      // If quirks.minWidth is defined, and the segment width is less than that
      // value, then render a left-aligned asset at minimum width boundary,
      // don't reposition it along the segment's left edge.
      // Note: this doesn't take into account the sprite.offsetX because no
      // asset with quirks.minWidth has offsetX defined right now.
      if (minWidthQuirk && actualWidth < minWidthQuirk) {
        x = 0
      } else {
        // This is the normal render logic.
        x =
          0 +
          (-left + (sprite.offsetX ?? 0) / TILE_SIZE_ACTUAL) *
            TILE_SIZE *
            multiplier
      }

      const distanceFromGround =
        multiplier *
        TILE_SIZE *
        ((svg.height - (sprite.originY ?? 0) + (sprite.offsetY ?? 0)) /
          TILE_SIZE_ACTUAL)

      drawSegmentImage(
        sprite.id,
        ctx,
        undefined,
        undefined,
        undefined,
        undefined,
        offsetLeft + x,
        groundLevel - distanceFromGround,
        undefined,
        undefined,
        multiplier,
        dpi
      )
    }
  }

  if (graphics.right) {
    const sprites = Array.isArray(graphics.right)
      ? graphics.right
      : [graphics.right]
    for (let l = 0; l < sprites.length; l++) {
      const sprite = getSpriteDef(sprites[l])
      const svg = images.get(sprite.id)

      // Skip drawing if sprite is missing
      if (!svg) continue

      let x
      // If quirks.minWidth is defined, and the segment width is less than that
      // value, then render a right-aligned asset at minimum width boundary,
      // don't reposition it along the segment's right edge.
      // Note: this doesn't take into account the sprite.offsetX because no
      // asset with quirks.minWidth has offsetX defined right now.
      if (minWidthQuirk && actualWidth < minWidthQuirk) {
        x =
          (minWidthQuirk - svg.width / TILE_SIZE_ACTUAL) *
          TILE_SIZE *
          multiplier
      } else {
        // This is the normal render logic.
        x =
          (-left +
            actualWidth -
            svg.width / TILE_SIZE_ACTUAL -
            (sprite.offsetX ?? 0) / TILE_SIZE_ACTUAL) *
          TILE_SIZE *
          multiplier
      }

      const distanceFromGround =
        multiplier *
        TILE_SIZE *
        ((svg.height - (sprite.originY ?? 0) + (sprite.offsetY ?? 0)) /
          TILE_SIZE_ACTUAL)

      drawSegmentImage(
        sprite.id,
        ctx,
        undefined,
        undefined,
        undefined,
        undefined,
        offsetLeft + x,
        groundLevel - distanceFromGround,
        undefined,
        undefined,
        multiplier,
        dpi
      )
    }
  }

  if (graphics.center) {
    const sprites = Array.isArray(graphics.center)
      ? graphics.center
      : [graphics.center]
    for (let l = 0; l < sprites.length; l++) {
      const sprite = getSpriteDef(sprites[l])
      const svg = images.get(sprite.id)

      // Skip drawing if sprite is missing
      if (!svg) continue

      const center = dimensions.center
      const offsetByPercentage =
        sprite.offsetX &&
        typeof sprite.offsetX === 'string' &&
        sprite.offsetX.endsWith('%')
      const offsetX = offsetByPercentage ? 0 : (sprite.offsetX ?? 0)
      const x =
        (center -
          svg.width / TILE_SIZE_ACTUAL / 2 -
          left -
          (offsetByPercentage ? percentToNumber(sprite.offsetX) * center : 0) -
          offsetX / TILE_SIZE_ACTUAL) *
        TILE_SIZE *
        multiplier
      const distanceFromGround =
        multiplier *
        TILE_SIZE *
        ((svg.height - (sprite.originY ?? 0) + (sprite.offsetY ?? 0)) /
          TILE_SIZE_ACTUAL)

      drawSegmentImage(
        sprite.id,
        ctx,
        undefined,
        undefined,
        undefined,
        undefined,
        offsetLeft + x,
        groundLevel - distanceFromGround,
        undefined,
        undefined,
        multiplier,
        dpi
      )
    }
  }

  // Draw items that are scattered from a pool of assets
  // Only used for random people generation right now
  if (graphics.scatter) {
    if (graphics.scatter.pool === 'people') {
      const offsetTop =
        (graphics.scatter.originY ??
          graphics.scatter.originY + PERSON_SPRITE_OFFSET_Y) ||
        PERSON_SPRITE_OFFSET_Y

      const people = PEOPLE
        // Temporarily: filter out all people that uses the `beach` tag
        // outside of Coastmix mode
        // TODO: figure out how to specify tags for inclusion/exclusion
        // possibly append the beach list when needed.
        .filter((person) => {
          if (coastmixMode) {
            return true
          } else {
            return !person.tags?.includes('beach')
          }
        })

      drawScatteredSprites(
        people,
        ctx,
        actualWidth,
        offsetLeft - left * TILE_SIZE * multiplier,
        offsetTop,
        groundLevel,
        randSeed,
        graphics.scatter.minSpacing,
        graphics.scatter.maxSpacing,
        PERSON_SPACING_ADJUSTMENT,
        graphics.scatter.padding,
        multiplier,
        dpi
      )
    }

    if (graphics.scatter.sprites) {
      drawScatteredSprites(
        graphics.scatter.sprites,
        ctx,
        actualWidth,
        offsetLeft - left * TILE_SIZE * multiplier,
        null,
        groundLevel,
        randSeed,
        graphics.scatter.minSpacing,
        graphics.scatter.maxSpacing,
        0,
        graphics.scatter.padding,
        multiplier,
        dpi
      )
    }
  }
}

export function getLocaleSegmentName (
  type: string,
  variantString: string
): string {
  const segmentInfo = getSegmentInfo(type)
  const variantInfo = getSegmentVariantInfo(type, variantString)
  const defaultName = variantInfo.name ?? segmentInfo.name
  const nameKey = variantInfo.nameKey ?? segmentInfo.nameKey
  const key = `segments.${nameKey}`

  return formatMessage(key, defaultName, { ns: 'segment-info' })
}

/**
 * TODO: remove this
 */
export function segmentsChanged (): void {
  const street = store.getState().street
  const updatedStreet = recalculateWidth(street)

  store.dispatch(
    updateSegments(
      updatedStreet.segments,
      updatedStreet.occupiedWidth,
      updatedStreet.remainingWidth
    )
  )

  saveStreetToServerIfNecessary()
}

/**
 * Process / sanitize segment labels
 *
 * @params name - Segment label to check
 * @returns normalized / sanitized segment label
 */
function normalizeSegmentLabel (label: string): string | undefined {
  label = label.trim()

  // If label is the empty string, return undefined
  if (label === '') {
    return undefined
  }

  // Trim a long label
  if (label.length > MAX_SEGMENT_LABEL_LENGTH) {
    label = label.substr(0, MAX_SEGMENT_LABEL_LENGTH) + '…'
  }

  return label
}

/**
 * Uses browser prompt to change the segment label
 *
 * @param segment - object describing the segment to edit
 * @param position - index of segment to edit
 */
export function editSegmentLabel (segment: Segment, position: number): void {
  const prevLabel =
    segment.label || getLocaleSegmentName(segment.type, segment.variantString)

  // If prompt returns empty string, set label to undefined. This resets the
  // label to the original default name
  // If prompt returns null (the prompt has been canceled), do not change the
  // label.
  const labelInput = window.prompt(
    formatMessage('prompt.segment-label', 'New segment label:'),
    prevLabel
  )

  if (labelInput === null) {
    return
  }

  const label = normalizeSegmentLabel(labelInput)

  if (label !== prevLabel) {
    store.dispatch(changeSegmentProperties(position, { label }))
    segmentsChanged()
  }
}

/**
 * Given the position of a segment or building, retrieve a reference to its
 * DOM element.
 *
 * @param position - either "left" or "right" for building,
 *              or a number for the position of the segment. Should be
 *              the `segmentIndex` or `position` variables.
 */
export function getSegmentEl (position: number | BoundaryPosition): HTMLElement {
  let segmentEl: HTMLElement

  if (position === BUILDING_LEFT_POSITION) {
    segmentEl = document.querySelectorAll(
      '.street-section-boundary'
    )[0] as HTMLElement
  } else if (position === BUILDING_RIGHT_POSITION) {
    segmentEl = document.querySelectorAll(
      '.street-section-boundary'
    )[1] as HTMLElement
  } else {
    const segments = document
      .getElementById('street-section-editable')!
      .querySelectorAll('.segment') as NodeListOf<HTMLElement>
    segmentEl = segments?.[position]
  }

  return segmentEl
}
