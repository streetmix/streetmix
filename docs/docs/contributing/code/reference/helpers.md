---
sidebar_position: 2
---

# Helpers and utilities

:::caution Under construction

This page is a work in progress.

:::

We use the following utilities to solve common programming problems.

## Pseudo-random number generator

If you need a one-time use random number (which will never be stored), there's nothing wrong with JavaScript's built-in `Math.random()`. However, there are times where we need a **seed** for a pseudo-random number generator (PRNG). Seeds generate a random-_looking_ sequence of numbers, but they have the benefit of remaining consistent across renders. For example, this is useful on our sidewalk segment, where we want to generate a random set of pedestrians but we want those pedestrians to remain the same between page refreshes or React component updates.

We can generate seeds by calling `generateRandSeed()`, then by feeding those seeds into the `seedrandom` library:

```js title="random-example.js"
import seedrandom from "seedrandom";
import { generateRandSeed } from "./util/random";

function doThingsWithThePRNG() {
  const seed = generateRandSeed();
  const prng1 = seedrandom(seed);

  // Generate pseudo-random numbers. Each time rng() is called, it returns
  // the next pseudo-random number in the sequence.
  console.log(prng1()); // -> first pseudo-random number
  console.log(prng1()); // -> second pseudo-random number
  // .. etc

  // Create another generator with the same seed
  const prng2 = seedrandom(seed);
  console.log(prng2()); // -> same value as the first call to prng1()
}
```

When `seedrandom()` is called, it returns a function that will return a consistent sequence of numbers each time it is called, for the given `seed`. The sequence returned is local to that function.

:::note

Although the `seed` can be any type of value, like a string, `generateRandSeed()` returns only integer values so that when we store seeds in a data model we can check for type consistency.

:::

In the above example, we call `seedrandom()` twice, with the same `seed`, so that we can create the same sequence of pseudo-random numbers in two different scenarios. This example is contrived because the pseudo-random number generators are used in the same function, and in reality the `seedrandom()` call may happen in separate parts of the application.

**References**

- [Random Seeds, Coded Hints, and Quintillions](http://davidbau.com/archives/2010/01/30/random_seeds_coded_hints_and_quintillions.html) [David Bau]
- [seedrandom.js](https://github.com/davidbau/seedrandom) [GitHub]

## axios

The [Axios](https://github.com/axios/axios) library was introduced as a recommendation over native `fetch` as a way to manage requests to the Streetmix back-end API layer. Additionally, Axios Mock Adapter offers the possibility to mock all API requests for integration testing.

If you work on API-related features, consider refactoring the `fetch` calls to use the API in `.assets/scripts/util/api.js`.
