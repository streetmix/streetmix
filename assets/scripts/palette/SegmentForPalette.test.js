/* eslint-env jest */
import React from 'react'
import { render } from '../../../test/helpers/render'
import { getVariantInfoDimensions } from '../segments/view'
import SegmentForPalette from './SegmentForPalette'

jest.mock('../segments/view')

describe('SegmentForPalette', () => {
  it('renders width correctly depending on the dimension', () => {
    const dimensions = { left: 100, right: 200 }
    getVariantInfoDimensions.mockImplementation(() => dimensions)

    const segment = {
      id: 'streetcar',
      name: 'Streetcar',
      nameKey: 'streetcar'
    }
    const { asFragment } = render(
      <SegmentForPalette segment={segment} randSeed={42} />
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
