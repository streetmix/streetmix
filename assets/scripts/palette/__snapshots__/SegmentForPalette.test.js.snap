// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentForPalette renders width correctly depending on the dimension 1`] = `
<DocumentFragment>
  <li
    class="segment segment-in-palette"
  >
    <div
      tabindex="0"
    >
      <img
        class="segment-image"
      />
    </div>
  </li>
</DocumentFragment>
`;
